File: MotoActions/src/org/lineageos/settings/device/actions/ChopForTorch.java
/*
 * SPDX-FileCopyrightText: 2015-2016 The CyanogenMod Project
 * SPDX-FileCopyrightText: The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.settings.device.actions;

import android.content.Context;
import android.hardware.camera2.CameraAccessException;
import android.hardware.camera2.CameraCharacteristics;
import android.hardware.camera2.CameraManager;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.os.VibrationEffect;
import android.os.Vibrator;
import android.util.Log;

import org.lineageos.settings.device.MotoActionsSettings;
import org.lineageos.settings.device.SensorHelper;

public class ChopForTorch implements UpdatedStateNotifier {
    private static final String TAG = "MotoActions-ChopForTorch";

    private final CameraManager mCameraManager;
    private final Vibrator mVibrator;
    private final MotoActionsSettings mMotoActionsSettings;
    private final SensorHelper mSensorHelper;
    private final Sensor mChopChopSensor;
    private final Sensor mProximitySensor;

    private String mRearCameraId;
    private boolean mIsEnabled;
    private boolean mProxIsCovered;
    private boolean mTorchEnabled;

    private CameraManager.TorchCallback mTorchCallback =
            new CameraManager.TorchCallback() {
        @Override
        public void onTorchModeChanged(String cameraId, boolean enabled) {
            if (!cameraId.equals(mRearCameraId)) {
                return;
            }
            mTorchEnabled = enabled;
        }

        @Override
        public void onTorchModeUnavailable(String cameraId) {
            if (!cameraId.equals(mRearCameraId)) {
                return;
            }
            mTorchEnabled = false;
        }
    };

    public ChopForTorch(MotoActionsSettings motoActionsSettings, Context context,
                             SensorHelper sensorHelper) {
        mMotoActionsSettings = motoActionsSettings;
        mSensorHelper = sensorHelper;
        mChopChopSensor = sensorHelper.getChopChopSensor();
        mProximitySensor = sensorHelper.getProximitySensor();
        mCameraManager = context.getSystemService(CameraManager.class);
        mCameraManager.registerTorchCallback(mTorchCallback, null);
        mVibrator = context.getSystemService(Vibrator.class);
    }

    @Override
    public synchronized void updateState() {
        if (mMotoActionsSettings.isChopChopGestureEnabled() && !mIsEnabled) {
            Log.d(TAG, "Enabling");
            mSensorHelper.registerListener(mChopChopSensor, mChopChopListener);
            mSensorHelper.registerListener(mProximitySensor, mProximityListener);
            mIsEnabled = true;
        } else if (!mMotoActionsSettings.isChopChopGestureEnabled() && mIsEnabled) {
            Log.d(TAG, "Disabling");
            mSensorHelper.unregisterListener(mChopChopListener);
            mSensorHelper.unregisterListener(mProximityListener);
            mIsEnabled = false;
        }
    }

    private final SensorEventListener mChopChopListener = new SensorEventListener() {
        @Override
        public synchronized void onSensorChanged(SensorEvent event) {
            Log.d(TAG, "chop chop triggered");
            if (mProxIsCovered) {
                Log.d(TAG, "proximity sensor covered, ignoring chop-chop");
                return;
            }

            if (mRearCameraId != null) {
                return;
            }

            try {
                mCameraManager.setTorchMode(mRearCameraId, !mTorchEnabled);
                mTorchEnabled = !mTorchEnabled;
            } catch (CameraAccessException ignored) {
            }
            mVibrator.vibrate(VibrationEffect.createOneShot(250, VibrationEffect.DEFAULT_AMPLITUDE));
        }

        @Override
        public void onAccuracyChanged(Sensor sensor, int accuracy) {
        }
    };

    private final SensorEventListener mProximityListener = new SensorEventListener() {
        @Override
        public synchronized void onSensorChanged(SensorEvent event) {
            float maxRange = Math.round(mProximitySensor.getMaximumRange() * 10f) / 10f;
            mProxIsCovered = event.values[0] < maxRange;
        }

        @Override
        public void onAccuracyChanged(Sensor sensor, int accuracy) {
        }
    };
}

File: packages/MotoActions/src/org/lineageos/settings/device/actions/ChopForTorch.kt
/*
 * SPDX-FileCopyrightText: 2015-2016 The CyanogenMod Project
 * SPDX-FileCopyrightText: The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.settings.device.actions

import android.content.Context
import android.content.SharedPreferences
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.camera2.CameraCharacteristics
import android.hardware.camera2.CameraManager
import android.hardware.camera2.CameraManager.TorchCallback
import android.os.VibrationEffect
import android.os.Vibrator
import android.util.Log
import org.lineageos.settings.device.MotoActionsSettings.GESTURE_CHOP_CHOP_KEY
import org.lineageos.settings.device.SensorHelper
import org.lineageos.settings.device.SensorHelper.Companion.SENSOR_TYPE_MMI_CHOP_CHOP

class ChopForTorch(
    context: Context,
    private val sharedPreferences: SharedPreferences,
    private val sensorHelper: SensorHelper,
) : UpdatedStateNotifier {

    private val cameraManager =
        context.getSystemService(CameraManager::class.java).apply {
            val callback =
                object : TorchCallback() {
                    override fun onTorchModeChanged(cameraId: String, enabled: Boolean) {
                        if (cameraId != rearCameraId) return
                        torchEnabled = enabled
                    }

                    override fun onTorchModeUnavailable(cameraId: String) {
                        if (cameraId != rearCameraId) return
                        torchEnabled = false
                    }
                }
            registerTorchCallback(callback, null)
            runCatching {
                cameraIdList.forEach { cameraId ->
                    val characteristics = getCameraCharacteristics(cameraId)
                    val orientation = characteristics.get(CameraCharacteristics.LENS_FACING)
                    if (orientation == CameraCharacteristics.LENS_FACING_BACK) {
                        rearCameraId = cameraId
                        return@forEach
                    }
                }
            }
        }
    private val vibrator = context.getSystemService(Vibrator::class.java)
    private val chopChopSensor: Sensor = sensorHelper.getSensor(SENSOR_TYPE_MMI_CHOP_CHOP)!!
    private val proximitySensor: Sensor = sensorHelper.getSensor(Sensor.TYPE_PROXIMITY)!!

    private var isEnabled = false
    private var proxIsCovered = false
    private var torchEnabled = false
    private var rearCameraId: String? = null

    @Synchronized
    override fun updateState() {
        val enabled = sharedPreferences.getBoolean(GESTURE_CHOP_CHOP_KEY, true)
        if (enabled == isEnabled) {
            return
        }
        if (enabled) {
            Log.d(TAG, "Enabling")
            sensorHelper.registerListener(chopChopSensor, chopChopListener)
            sensorHelper.registerListener(proximitySensor, proximityListener)
        } else {
            Log.d(TAG, "Disabling")
            sensorHelper.unregisterListener(chopChopListener)
            sensorHelper.unregisterListener(proximityListener)
        }
        isEnabled = enabled
    }

    private val chopChopListener =
        object : SensorEventListener {
            @Synchronized
            override fun onSensorChanged(event: SensorEvent) {
                Log.d(TAG, "chop chop triggered")
                if (proxIsCovered) {
                    Log.d(TAG, "proximity sensor covered, ignoring chop-chop")
                    return
                }
                vibrator.vibrate(
                    VibrationEffect.createOneShot(250, VibrationEffect.DEFAULT_AMPLITUDE)
                )

                rearCameraId?.let { id ->
                    runCatching {
                        cameraManager.setTorchMode(id, !torchEnabled)
                        torchEnabled = !torchEnabled
                    }
                }
            }

            override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}
        }

    private val proximityListener =
        object : SensorEventListener {
            @Synchronized
            override fun onSensorChanged(event: SensorEvent) {
                val maxRange = kotlin.math.round(proximitySensor.maximumRange * 10f) / 10f
                proxIsCovered = event.values[0] < maxRange
            }

            override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}
        }

    companion object {
        private const val TAG = "ChopForTorch"
    }
}
